DAOS-3492 control: enable daos_server to live without instances

tanabarr opened the pull request:
Enable the daos_server to continue running when the instances are
terminated. Record running state on the instance cmd runner.

- add atomic IsStarted() for instance runner
- don't quit harness.Start() on instance runner stop
- tabulate host groups using a generic function
- clarify intent with better code organisation and comments

Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1523/2/execution/node/516/log

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-1523/2/display/redirect

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1523/3/execution/node/554/log

tanabarr commented:
Example output when querying members then issuing "dmg system stop" command.
Member instances are stopped but daos_server processes remain active and responsive to dmg commands.
```
[tanabarr@wolf-71 ~]$ dmg -l wolf-71:10001 system member-query -i
wolf-71:10001: connected
System-member-query command SUCCEEDED: 0.0.0.0:10001/4dc6b6b4-f518-490e-b395-9679d629f553/0 10.8.1.72:10001/39d6216f-ec31-44c4-99ac-a4d809ea0ef2/1
[tanabarr@wolf-71 ~]$ dmg -l wolf-71:10001 system stop -i
wolf-71:10001: connected
System-stop command SUCCEEDED: : still 2 active members
[tanabarr@wolf-71 ~]$ dmg -l wolf-71:10001 system member-query -i
wolf-71:10001: connected
System-member-query command SUCCEEDED: no joined members
[tanabarr@wolf-71 ~]$ dmg -l wolf-7[1-2]:10001 storage scan -i --summary
wolf-[71-72]:10001: connected

 HOSTS          SCM                     NVME
 -----          ---                     ----
 wolf-[71-72]   5.79TB (2 namespaces)   3.03TB (3 controllers)
```

mjmac commented on src/control/server/harness.go:
```diff
@@ -240,14 +243,27 @@ func (h *IOServerHarness) Start(parent context.Context) error {
+		if err != nil {
 			return err
 		}
+		if newMember != nil {
```
This is a little confusing. After looking at the changes to StartManagementService(), I guess that newMember can be nil if this method is called on a non-MS instance, right? Generally, using nil as a valid return value isn't desirable because it's atypical and requires a greater level of care in its use.

At a minimum, I'd suggest renaming the variable to something like msReplicaMember to make it clear that it's special, as well as checking for the nil case and returning early (along with a comment explaining this) so that you can outdent the rest of the logic. Even better would probably be to refactor things a bit to avoid calling StartManagementService() unless we're reasonably certain that it should succeed.

kjacque replied:
Is it possible that newMember will be nil even if there was no error?

mjmac commented on src/control/server/harness.go:
```diff
@@ -230,7 +233,7 @@ func (h *IOServerHarness) Start(parent context.Context) error {
 		// TODO: Restart failed instances rather than shutting everything
 		// down.
-		h.log.Errorf("instance error: %s", err)
+		h.log.Errorf("instance %d error: %s", err)
```
Missing Index() ?

mjmac commented on src/control/server/ctl_system.go:
```diff
@@ -83,15 +47,74 @@ func (svc *ControlService) SystemMemberQuery(ctx context.Context, req *ctlpb.Sys
+
+	// TODO: inhibit rebuild on pool services, parallelise and make async.
+	for _, rank := range members.GetRanks() {
+		member := members[rank]
```
This is an unsafe map access. Wouldn't it be better to iterate over the slice returned by members.GetMembers() ?

kjacque replied:
Wondering if we could just get the member via "range members"?

mjmac commented on src/control/common/system.go:
```diff
@@ -91,60 +133,49 @@ func (smrs SystemMemberResults) String() string {
 }
 
-// GetMembers returns internal member structs as a sequence.
+// GetMembers returns SystemMembers.
```
This change removes all thread safety. Not sure if that was intended, but now callers of this method will have free access to the previously-protected private membership map.

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1523/6/execution/node/555/log

tanabarr commented:
> The system-member query output is difficult to read and won't scale well. Also, the SUCCEEDED messages seem to be more appropriate as debug-level messages or removed entirely.
> 
> I'd also suggest moving the membership stuff out of common -- that seems like an odd home for it given that it's only used by the server package, right? Perhaps it could move into the server package or a new package under server, e.g. server/membership.

have addressed most of these comments in https://github.com/daos-stack/daos/pull/1585

tanabarr commented:
> I'm curious if you've tested what happens if you shoot a daos_io_server instance with a signal, instead of using the safe shutdown.

It is a very similar situation to using the safe shutdown, `[tanabarr@wolf-71 daos_m]$ ssh root@wolf-72 "killall daos_io_server"` yields the same situation where instances have been stopped but the harness is still running.

The reporting of errors/exit state from terminated `daos_io_server` instances needs more thought. Currently I just log the exit message to ERROR.

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1523/8/execution/node/485/log

kjacque commented on src/control/server/harness.go:
```diff
@@ -244,17 +244,22 @@ func (h *IOServerHarness) Start(parent context.Context, membership *system.Membe
+			h.log.Errorf("instance exited: %v", err)
+			//			if err != nil {
+			//				return errors.Wrap(err, "Instance error")
+			//			}
```
Should this be removed? Looks like instead of killing all instances (as noted above in the comment) now we're just logging some debug output.

kjacque commented on src/control/server/ioserver/exec.go:
```diff
@@ -50,8 +51,9 @@ type (
-		log    logging.Logger
+		Config  *Config
+		log     logging.Logger
+		started uint32
```
I guess this is a uint32 specifically to take advantage of the atomic library. It was counterintuitive at first - since it is used as a bool. The goal is to avoid using mutexes, I'm guessing?

tanabarr replied:
yes for the simple use case I think a mutex is unnecessary

daosbuild1 commented:
Test stage run_test.sh completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1523/9/execution/node/348/log

kjacque commented on src/control/server/harness.go:
```diff
@@ -36,16 +37,20 @@ import (
-		log:       log,
-		instances: make([]*IOServerInstance, 0, 2),
+		log:            log,
+		instances:      make([]*IOServerInstance, 0, 2),
```
No need to make a change in this PR, but just wondering - is there a const defined somewhere for max number of instances?

tanabarr replied:
I doubt it given @mjmac hardcoded it here but I'm sure we can create one

mjmac replied:
There is a constant defined at https://github.com/daos-stack/daos/blob/master/src/control/server/server.go#L59 ... Wouldn't hurt to use that here. Remember though that this value is just a starting point rather than a hard cap on the slice length.

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1523/10/execution/node/503/log

tanabarr commented:
Example command output after updates related to feedback from @johannlombardi and @mjmac :
```
[tanabarr@wolf-71 daos_m]$ dmg -i system member-query
localhost:10001: connected
Rank    UUID                                    Control Address State
----    ----                                    --------------- -----
0       ea41d684-62fd-42ca-9db4-2dd801c03425    10.8.1.71:10001 Started
1       070ace84-1459-4386-9c35-c4440ce2d44f    10.8.1.72:10001 Started
[tanabarr@wolf-71 daos_m]$ dmg -i system stop  --prep
localhost:10001: connected
Ranks           Operation       Result
-----           ---------       ------
rank[0-1]       prep shutdown   OK
[tanabarr@wolf-71 daos_m]$ dmg -i system member-query
localhost:10001: connected
Rank    UUID                                    Control Address State
----    ----                                    --------------- -----
0       ea41d684-62fd-42ca-9db4-2dd801c03425    10.8.1.71:10001 Stopping
1       070ace84-1459-4386-9c35-c4440ce2d44f    10.8.1.72:10001 Stopping
[tanabarr@wolf-71 daos_m]$ dmg -i system stop  --kill
localhost:10001: connected
Ranks           Operation       Result
-----           ---------       ------
rank[0-1]       stop            OK
[tanabarr@wolf-71 daos_m]$ dmg -i system member-query
localhost:10001: connected
No members in system
[tanabarr@wolf-71 daos_m]$
```

daosbuild1 commented:
Test stage Functional completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1523/14/execution/node/485/log

kjacque commented on src/control/server/harness.go:
```diff
@@ -260,39 +252,57 @@ func (h *IOServerHarness) Start(parent context.Context, membership *system.Membe
+				if h.HasStartedInstances() {
+					return errors.New("attempts to restart when instances are already running")
+				}
+				atomic.StoreInt32(&h.readyForRestart, 0)
```
Realized after spending some time poring over this that it doesn't seem anyone is actually reading this readyForRestart boolean. If I'm being honest, it's a little hard to follow the expected interaction among the threads. It would be really nice to have some kind of diagram or flowchart of how this all works together, and the actual flow of each method that gets us to each of these cases.

tanabarr replied:
the readyForRestart is going to be used in the start logic in https://github.com/daos-stack/daos/pull/1541 . will add some diagram/detail to the ticket.

tanabarr commented on src/control/cmd/dmg/utils.go:
```diff
@@ -146,3 +147,24 @@ func formatHostGroupResults(buf *bytes.Buffer, groups hostlist.HostGroups) strin
 }
+
+// groupSummaryTable is a helper function that prints hostgroups with 2 column entries
+func groupSummaryTable(firstTitle, secondTitle, thirdTitle string, groups hostlist.HostGroups) (string, error) {
```
sounds good, will make the adjustments

mjmac replied:
Agree with Kris on this. You could also move the groups parameter to be first, and then make titles a variadic string param for maximum flexibility.

tanabarr commented on src/control/server/harness.go:
```diff
@@ -260,39 +252,57 @@ func (h *IOServerHarness) Start(parent context.Context, membership *system.Membe
+				msg := fmt.Sprintf("instance exited: %v", err)
+				if allInstancesStopped {
+					msg += ", all instances stopped!"
+					atomic.StoreInt32(&h.readyForRestart, 1)
```
was going to do as a follow-up but I will do it here

mjmac commented on src/control/server/harness.go:
```diff
@@ -256,20 +254,96 @@ func (h *IOServerHarness) Start(parent context.Context, membership *system.Membe
+		if err := h.waitInstancesReady(ctx); err != nil {
+			return err
+		}
+		if err := h.monitorInstances(ctx); err != nil {
```
I think it'd be worth adding a comment to explain the logic here... I understand it now but I had to study the code. In particular, it's not obvious that sending a restart signal after all instances have stopped results in another loop iteration.

tanabarr replied:
have added comments

mjmac commented on src/control/server/harness.go:
```diff
@@ -256,20 +254,96 @@ func (h *IOServerHarness) Start(parent context.Context, membership *system.Membe
+	h.restartInstances <- struct{}{} // trigger harness to restart its instances
+
 	return nil
 }
```
I'm a little concerned that all of this logic isn't being done under a single lock. Given that there could potentially be multiple restart messages received in close succession, it's possible that you could get some undefined behavior if one request starts to change state while another request is being processed at nearly the same time.

tanabarr replied:
this function has been moved to https://github.com/daos-stack/daos/pull/1541 and lock added there

mjmac commented on src/control/server/harness.go:
```diff
@@ -256,20 +254,96 @@ func (h *IOServerHarness) Start(parent context.Context, membership *system.Membe
+		return errors.New("can't start instances: already started")
+	}
+
+	h.restartInstances <- struct{}{} // trigger harness to restart its instances
```
How will this method be called? If it's in the same goroutine as the one running monitorInstances(), this could result in a deadlock. harness.Start() is being called in a new goroutine from server.go so this is probably OK, but just something to be careful with.

tanabarr replied:
moved to https://github.com/daos-stack/daos/pull/1541, will never be the same goroutine as running monitor() because it's only ever called over gRPC in a handler, not by the goroutine spawned from server main thread itself (running Start()).

mjmac commented on src/control/system/member.go:
```diff
@@ -83,14 +83,14 @@ type Members []*Member
 // its string representation "address/rank".
 type MemberResult struct {
-	ID     string
+	Rank   uint32
```
I wonder if it makes sense to make this an ioserver.Rank instead, to avoid all of the casting. I don't feel strongly about it, just something that occurred to me.

tanabarr replied:
you still have to then cast in and out of Protobuf messages

tanabarr commented on src/control/server/harness.go:
```diff
@@ -36,16 +37,20 @@ import (
+		log:              log,
+		instances:        make([]*IOServerInstance, 0, 2),
+		restartInstances: make(chan struct{}, 1),
+		errChan:          make(chan error, 2),
```
using the constant defined in server.go

tanabarr commented on src/control/server/harness.go:
```diff
@@ -36,16 +37,20 @@ import (
-		instances: make([]*IOServerInstance, 0, 2),
+		log:              log,
+		instances:        make([]*IOServerInstance, 0, 2),
+		restartInstances: make(chan struct{}, 1),
```
yes but I don't want to close that channel to unblock it in this situation, because it will be reused

tanabarr commented on src/control/server/harness.go:
```diff
@@ -196,43 +203,36 @@ func registerNewMember(membership *system.Membership, instance *IOServerInstance
-		// If this instance is a MS replica, register membership
+
 		if instance.IsMSReplica() {
 			if err := registerNewMember(membership, instance); err != nil {
```
this will be addressed in the restart PR (https://github.com/daos-stack/daos/pull/1541)

tanabarr commented:
gatekeeper please use PR description as commit message

kjacque commented on src/control/server/server.go:
```diff
@@ -201,18 +201,10 @@ func Start(log *logging.LeveledLogger, cfg *Configuration) error {
+	}()
 
-	return errors.Wrap(harness.Start(ctx, membership), "DAOS I/O Server exited with error")
+	return errors.Wrap(<-harnessErr, "DAOS harness exited with error")
```
No change requested, but I'm curious how this approach changes the flow from making the call synchronously. I don't really see the difference here.

tanabarr replied:
it's not strictly necessary and there isn't much functional difference but at some point I'm anticipating wanting to handle messages/errors from the harness in main. that decoupling would facilitate a model where the harness can be interacted with by both the main server and gRPC handlers. so whilst there is no functional difference now it's a step towards that decoupling which would mean the server could still operate even if there were problems with the harness (as it doesn't add current value I will revert if you want).

mjmac replied:
I have a different concern about this... It's generally preferable to handle the concurrency inside the concurrent method rather than just throwing a goroutine around a method that may or may not have been designed for concurrency. In other words, if the guts of harness.Start() are intended to be run in a new goroutine, then the method should be adjusted to reflect this, along with handling all of the control/error stuff. You could make harness.Start() accept an error channel to be monitored by the main goroutine, similar to the pattern that we use for starting individual instances.

That way if there is an error in setting things up prior to launching the new goroutine, Start() can exit and return an error synchronously; otherwise you can just block on the error channel after normal startup.

kjacque replied:
@tanabarr I'm fine with either reverting for now, or taking the approach @mjmac talks about.

tanabarr replied:
will revert and address when needed

daosbuild1 commented:
Test stage Functional_Hardware completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-1523/18/execution/node/555/log

mjmac commented on src/control/cmd/dmg/utils.go:
```diff
@@ -146,3 +147,28 @@ func formatHostGroupResults(buf *bytes.Buffer, groups hostlist.HostGroups) strin
+func tabulateHostGroups(groupTitle string, columnTitles []string, groups hostlist.HostGroups) (string, error) {
+	titles := []string{groupTitle}
+	titles = append(titles, columnTitles...)
+
```
This could be simpler if you make the titles param variadic:

```go
func tabulateHostGroups(groups hostlist.HostGroups, titles ...string) (string, error) {
        if len(titles) < 2 {
                return "", errors.New("incorrect number of titles")
        }
...
```
